### ISSUE TITLE
Hi Paarth 

### ISSUE THREAD

**Issue by fajieyuan** [paarthneekhara/byteNet-tensorflow#10]
thanks for the great work. However, I noticed code  may have a serious bug.
I am now playing with train_generator.py but find that the code may have some problems.

In your original evaluatio, there is no training and testing set.

So if we divide the dataset by randomly splitting the whole data. Then we slightly change the code with a simple testing set loss evaluation which like training loss.

You may find that the loss of training is decreasing from the beginning but the loss in testing set never decrese but actually increase from the beggining.

It is kind of weird, do you know the reason, do you think there may be a problem with the algorithm?

Thanks

Fajie

**Comment by fajieyuan** [comment]
Hi， i think the problem is caused by my dataset. I used another data now, which has decreased loss with training.

**Comment by fajieyuan** [comment]
hi I am not sure about masked 1d, It seems that the mask operation is done on difference channels like image in RGB channel. Do you think your implementation here is in line with the paper, see pixelRNN

**Comment by paarthneekhara** [comment]
Hi! Masking implies masking out future tokens for conditioning the output at current timestep. It is implemented as a causal convolution as suggested in the paper. See ops.py conv1d implementation.

**Comment by fajieyuan** [comment]
Hi Paarth,

Thanks so much. I know the basic idea of masking i.e. ignoring the future tokens. However, I just checked conv1d but cannot find which line stands for mask. Based on my understanding, your masking step is done by using this line:
  source_sentence = self.t_sentence[:, 0:-1]
Am I right, this line means you only use the previous t-1 for training tokens but not the latest one, am I right?

I cannot see which line in conv1d using masking, sorry for the bothering you so much because I was confused by masking for quite a long time,

Thanks
Fajie

**Comment by paarthneekhara** [comment]
The argument causal in conv1d implementation can be set to True (for masking) or False for the encoder. The way masking is implemented is by pre-padding the input with 0s (see the code under if causal on how much to pad). Since the input sequence is now longer than the expected output sequence, we slice the output beyond the required number of tokens. This can be a little tricky to understand, but once you'll try to write it down on pen and paper, you'll see that the output at any timestep doesn't depend on future tokens in the input.

**Comment by fajieyuan** [comment]
thanks a lot, Now I got the idea,
WIsh you a happy new year,
Fajie
